src/rfq/protocols/bebop/tycho_decoder.rs

```diff
@@ -0,0 +1,228 @@
+use std::collections::HashMap;
+
+use tycho_client::feed::{synchronizer::ComponentWithState};
+use tycho_common::{models::token::Token, Bytes};
+
+use super::{models::BebopPriceData, state::BebopState};
+use crate::{
+    protocol::{errors::InvalidSnapshotError, models::TryFromWithBlock},
+    rfq::models::TimestampHeader,
+};
+
+impl TryFromWithBlock<ComponentWithState, TimestampHeader> for BebopState {
+    type Error = InvalidSnapshotError;
+
+    async fn try_from_with_block(
+        snapshot: ComponentWithState,
+        timestamp_header: TimestampHeader,
+        _account_balances: &HashMap<Bytes, HashMap<Bytes, Bytes>>,
+        all_tokens: &HashMap<Bytes, Token>,
+    ) -> Result<Self, Self::Error> {
+        let state_attrs = snapshot.state.attributes;
+
+        if snapshot.component.tokens.len() != 2 {
+            return Err(InvalidSnapshotError::ValueError(
+                "Component must have 2 tokens (base and quote)".to_string(),
+            ));
+        }
+
+        let base_token_address = snapshot.component.tokens[0].clone();
+        let quote_token_address = snapshot.component.tokens[1].clone();
+
+        let base_token = all_tokens
+            .get(&base_token_address)
+            .ok_or_else(|| {
+                InvalidSnapshotError::ValueError(format!(
+                    "Base token not found: {}",
+                    hex::encode(&base_token_address)
+                ))
+            })?
+            .clone();
+
+        let quote_token = all_tokens
+            .get(&quote_token_address)
+            .ok_or_else(|| {
+                InvalidSnapshotError::ValueError(format!(
+                    "Quote token not found: {}",
+                    hex::encode(&quote_token_address)
+                ))
+            })?
+            .clone();
+
+        let bids_json = state_attrs
+            .get("bids")
+            .ok_or_else(|| InvalidSnapshotError::MissingAttribute("bids".to_string()))?;
```

Comment:
I thought we agreed for the key to be missing no? or am I confused? if it's missing then the levels are empty

Reply:
removing this check